### PR TITLE
Crossdock echo behaviors test

### LINKED ISSUE #61: Implement Crossdock echo behaviors
parent: https://github.com/yarpc/yarpc/issues/17

### PR DESCRIPTION
**Pull request by junchaowu**
Resolves #61

- [x] implement echo behaviors
- [x] add `yarpc-go:crossdock-echo-behaviors`
- [x] add `yarpc-node:crossdock-echo-behaviors`
- [x] add `yarpc-python:crossdock-echo-behaviors`

### CODE REVIEW

**Comment by junchaowu** [comment]
```
yarpc-node_1 | {"log":"resetting connection","level":"info","error":{"type":"tchannel.socket","message":"tchannel socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082","hostPort":null,"direction":"out","remoteAddr":null,"name":"TchannelSocketError","socketRemoteAddr":"127.0.0.1:8082","causeMessage":"connect ECONNREFUSED 127.0.0.1:8082","origMessage":"connect ECONNREFUSED 127.0.0.1:8082","code":"ECONNREFUSED","errno":"ECONNREFUSED","syscall":"connect","fullType":"tchannel.socket~!~error.wrapped-io.connect.ECONNREFUSED"},"numInOps":0,"numOutOps":0,"inPending":0,"outPending":0,"hostPort":null,"channelListened":false,"channelListening":false,"channelDestroyed":false,"channelDraining":false,"connGUID":"6~","connDirection":"out","socketRemoteAddr":"127.0.0.1:8082","remoteName":null,"connClosing":true}
crossdock_1  | FAILED - {yarpc-node map[behavior:thrift transport:tchannel server:yarpc-node]} - TchannelSocketError: tchannel socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082
crossdock_1  |     at createError (/src/node_modules/error/typed.js:31:22)
crossdock_1  |     at Object.createError (/src/node_modules/error/wrapped.js:88:19)
crossdock_1  |     at TChannelConnection.onSocketError (/src/node_modules/tchannel/connection.js:903:30)
crossdock_1  |     at Socket.onSocketError (/src/node_modules/tchannel/connection.js:476:14)
crossdock_1  |     at emitOne (events.js:77:13)
crossdock_1  |     at Socket.emit (events.js:169:7)
crossdock_1  |     at emitErrorNT (net.js:1250:8)
crossdock_1  |     at nextTickCallbackWith2Args (node.js:442:9)
crossdock_1  |     at process._tickCallback (node.js:356:17)
crossdock_1  |
crossdock_1  | Tests did not pass!
crossdock_1  |
yarpc-node_1 | {"log":"YARPC inbound request error","level":"info","message":"tchannel socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082","type":"tchannel.socket","direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":3,"remainingms":97,"start":"2016-03-11T23:08:55.850Z","deadline":"2016-03-11T23:08:55.950Z"}
yarpcjava_crossdock_1 exited with code 1

socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082
crossdock_1  |     at createError (/src/node_modules/error/typed.js:31:22)
crossdock_1  |     at Object.createError (/src/node_modules/error/wrapped.js:88:19)
crossdock_1  |     at TChannelConnection.onSocketError (/src/node_modules/tchannel/connection.js:903:30)
crossdock_1  |     at Socket.onSocketError (/src/node_modules/tchannel/connection.js:476:14)
crossdock_1  |     at emitOne (events.js:77:13)
crossdock_1  |     at Socket.emit (events.js:169:7)
crossdock_1  |     at emitErrorNT (net.js:1250:8)
crossdock_1  |     at nextTickCallbackWith2Args (node.js:442:9)
crossdock_1  |     at process._tickCallback (node.js:356:17)
yarpc-node_1 | {"log":"YARPC outbound request","level":"info","direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":0,"remainingms":100,"start":"2016-03-11T23:08:55.825Z","deadline":"2016-03-11T23:08:55.925Z"}
yarpc-node_1 | {"log":"YARPC outbound request over HTTP","level":"debug","byteLength":19,"direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":null,"remainingms":null,"start":"2016-03-11T23:08:55.825Z","deadline":"2016-03-11T23:08:55.925Z"}
yarpc-node_1 | {"log":"YARPC inbound request","level":"info","direction":"inbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":null,"ttlms":100,"elapsedms":0,"remainingms":100,"start":"2016-03-11T23:08:55.835Z","deadline":"2016-03-11T23:08:55.935Z"}
yarpc-node_1 | {"log":"YARPC inbound request","level":"info","direction":"inbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":null,"ttlms":100,"elapsedms":0,"remainingms":100,"start":"2016-03-11T23:08:55.835Z","deadline":"2016-03-11T23:08:55.935Z"}
yarpc-node_1 | {"log":"YARPC outbound response","level":"info","direction":"inbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":null,"ttlms":100,"elapsedms":1,"remainingms":99,"start":"2016-03-11T23:08:55.835Z","deadline":"2016-03-11T23:08:55.935Z"}
crossdock_1  | PASSED - {yarpc-node map[transport:http server:yarpc-node behavior:thrift]} - Server said: 183g2b6
yarpc-node_1 | {"log":"YARPC inbound response over HTTP","level":"debug","byteLength":19,"direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":14,"remainingms":86,"start":"2016-03-11T23:08:55.825Z","deadline":"2016-03-11T23:08:55.925Z"}
yarpc-node_1 | {"log":"YARPC inbound response","level":"info","direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":14,"remainingms":86,"start":"2016-03-11T23:08:55.825Z","deadline":"2016-03-11T23:08:55.925Z"}
yarpc-node_1 | {"log":"YARPC outbound request","level":"info","direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":0,"remainingms":100,"start":"2016-03-11T23:08:55.850Z","deadline":"2016-03-11T23:08:55.950Z"}
yarpc-node_1 | {"log":"resetting connection","level":"info","error":{"type":"tchannel.socket","message":"tchannel socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082","hostPort":null,"direction":"out","remoteAddr":null,"name":"TchannelSocketError","socketRemoteAddr":"127.0.0.1:8082","causeMessage":"connect ECONNREFUSED 127.0.0.1:8082","origMessage":"connect ECONNREFUSED 127.0.0.1:8082","code":"ECONNREFUSED","errno":"ECONNREFUSED","syscall":"connect","fullType":"tchannel.socket~!~error.wrapped-io.connect.ECONNREFUSED"},"numInOps":0,"numOutOps":0,"inPending":0,"outPending":0,"hostPort":null,"channelListened":false,"channelListening":false,"channelDestroyed":false,"channelDraining":false,"connGUID":"6~","connDirection":"out","socketRemoteAddr":"127.0.0.1:8082","remoteName":null,"connClosing":true}
crossdock_1  | FAILED - {yarpc-node map[behavior:thrift transport:tchannel server:yarpc-node]} - TchannelSocketError: tchannel socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082
crossdock_1  |     at createError (/src/node_modules/error/typed.js:31:22)
crossdock_1  |     at Object.createError (/src/node_modules/error/wrapped.js:88:19)
crossdock_1  |     at TChannelConnection.onSocketError (/src/node_modules/tchannel/connection.js:903:30)
crossdock_1  |     at Socket.onSocketError (/src/node_modules/tchannel/connection.js:476:14)
crossdock_1  |     at emitOne (events.js:77:13)
crossdock_1  |     at Socket.emit (events.js:169:7)
crossdock_1  |     at emitErrorNT (net.js:1250:8)
crossdock_1  |     at nextTickCallbackWith2Args (node.js:442:9)
crossdock_1  |     at process._tickCallback (node.js:356:17)
crossdock_1  |
crossdock_1  | Tests did not pass!
crossdock_1  |
yarpc-node_1 | {"log":"YARPC inbound request error","level":"info","message":"tchannel socket error (ECONNREFUSED from connect): connect ECONNREFUSED 127.0.0.1:8082","type":"tchannel.socket","direction":"outbound","caller":"yarpc-test","service":"yarpc-test","procedure":"Echo::echo","encoding":"thrift","ttlms":100,"elapsedms":3,"remainingms":97,"start":"2016-03-11T23:08:55.850Z","deadline":"2016-03-11T23:08:55.950Z"}
yarpcjava_crossdock_1 exited with code 1
```

**Comment by junchaowu** [comment]
crossdock fails because node is using 127.0.0.1 to connect to java tchannel instance however it has to be realy IP to connect.

**Comment by breerly** [comment]
Looking at this now - sounds like a race of some kind.

**Comment by breerly** [comment]
I think that https://github.com/yarpc/crossdock/blob/master/execute/wait.go#L14 is where the problem is - I don't think `Wait` is prepared for this type of error...

**Comment by breerly** [comment]
Ok - managed to get some output https://travis-ci.org/yarpc/yarpc-java#L6670

```
yarpc-node_1 | 
yarpc-node_1 | assert.js:89
yarpc-node_1 |   throw new assert.AssertionError({
yarpc-node_1 |   ^
crossdock_1  | 
crossdock_1  | Crossdock starting...
crossdock_1  | 
yarpc-node_1 | AssertionError: Expected host to be a string
yarpc-node_1 |     at TChannel.listen (/src/node_modules/tchannel/channel.js:632:9)
yarpc-node_1 |     at TChannelInbound.start (/src/transport/tchannel/inbound.js:29:18)
yarpc-node_1 |     at RPC.start (/src/rpc.js:173:26)
yarpc-node_1 |     at App.start (/src/test/subject.js:218:14)
yarpc-node_1 |     at TestHarness.start (/src/test/harness.js:59:18)
yarpc-node_1 |     at main (/src/test/harness.js:145:13)
yarpc-node_1 |     at Object.<anonymous> (/src/test/harness.js:162:2)
yarpc-node_1 |     at Module._compile (module.js:409:26)
yarpc-node_1 |     at Object.Module._extensions..js (module.js:416:10)
yarpc-node_1 |     at Module.load (module.js:343:32)
yarpcjava_yarpc-node_1 exited with code 1
crossdock_1  | Waiting on CROSSDOCK_CLIENTS=[yarpc-java yarpc-node]
crossdock_1  | 
yarpc-java_1 | 23:06:22.312 [nioEventLoopGroup-5-0] DEBUG io.netty.util.Recycler - -Dio.netty.recycler.maxCapacity: 262144
crossdock_1  | HTTP: Service yarpc-java:8080 is up
crossdock_1  | 2016/03/14 23:06:52 HTTP: Service yarpc-node:8080 timed out. Last error: Head http://yarpc-node:8080/: dial tcp: lookup yarpc-node on 127.0.0.11:53: no such host
crossdock_1  | 2016/03/14 23:06:52 Error: One or more services timed out after 30000000000 second(s)
yarpcjava_crossdock_1 exited with code 1
```

**Comment by breerly** [comment]
Ok - Go also failed in the same way against node: https://travis-ci.org/yarpc/yarpc-go/jobs/116001255#L515

**Comment by breerly** [comment]
Tried to switch java to just call go, but it failed locally:

```
yarpc-java_1   | org.junit.ComparisonFailure: expected:<[internal service error: unknown procedure "echo/raw"
yarpc-java_1   | ]> but was:<[c5f4d19]>
yarpc-java_1   |        at org.junit.Assert.assertEquals(Assert.java:115)
yarpc-java_1   |        at org.junit.Assert.assertEquals(Assert.java:144)
yarpc-java_1   |        at io.yarpc.harness.behavior.raw.RawTest.testEcho(RawTest.java:43)
yarpc-java_1   |        at io.yarpc.harness.behavior.HarnessTest.testRaw(HarnessTest.java:144)
yarpc-java_1   |        at io.yarpc.harness.behavior.HarnessTest.behaviorTest(HarnessTest.java:114)
yarpc-java_1   |        at io.yarpc.crossdock.HarnessServerInboundHandler.harness(HarnessServerInboundHandler.java:89)
yarpc-java_1   |        at io.yarpc.crossdock.HarnessServerInboundHandler.messageReceived(HarnessServerInboundHandler.java:73)
yarpc-java_1   |        at io.yarpc.crossdock.HarnessServerInboundHandler.messageReceived(HarnessServerInboundHandler.java:41)
yarpc-java_1   |        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
yarpc-java_1   |        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:84)
yarpc-java_1   |        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153)
yarpc-java_1   |        at io.netty.channel.PausableChannelEventExecutor.invokeChannelRead(PausableChannelEventExecutor.java:86)
yarpc-java_1   |        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:389)
yarpc-java_1   |        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
yarpc-java_1   |        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:84)
yarpc-java_1   |        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153)
yarpc-java_1   |        at io.netty.channel.PausableChannelEventExecutor.invokeChannelRead(PausableChannelEventExecutor.java:86)
yarpc-java_1   |        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:389)
yarpc-java_1   |        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:243)
yarpc-java_1   |        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:84)
yarpc-java_1   |        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:153)
yarpc-java_1   |        at io.netty.channel.PausableChannelEventExecutor.invokeChannelRead(PausableChannelEventExecutor.java:86)
yarpc-java_1   |        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:389)
yarpc-java_1   |        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:956)
yarpc-java_1   |        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:127)
yarpc-java_1   |        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:514)
yarpc-java_1   |        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:471)
yarpc-java_1   |        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:385)
yarpc-java_1   |        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:351)
yarpc-java_1   |        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
yarpc-java_1   |        at io.netty.util.internal.chmv8.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1412)
yarpc-java_1   |        at io.netty.util.internal.chmv8.ForkJoinTask.doExec(ForkJoinTask.java:280)
yarpc-java_1   |        at io.netty.util.internal.chmv8.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:877)
yarpc-java_1   |        at io.netty.util.internal.chmv8.ForkJoinPool.scan(ForkJoinPool.java:1706)
yarpc-java_1   |        at io.netty.util.internal.chmv8.ForkJoinPool.runWorker(ForkJoinPool.java:1661)
yarpc-java_1   |        at io.netty.util.internal.chmv8.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:126)
```

**Comment by breerly** [comment]
Ok - I have Java running against itself just fine:

```

Crossdock starting...

Waiting on CROSSDOCK_CLIENTS=[yarpc-java]

HTTP: Service yarpc-java:8080 is up

All services are up after 384.653231ms!

Executing Matrix...

PASSED - {yarpc-java map[server:yarpc-java behavior:raw transport:http]} - b72dce0
PASSED - {yarpc-java map[server:yarpc-java behavior:raw transport:tchannel]} - a776c23
PASSED - {yarpc-java map[transport:http server:yarpc-java behavior:json]} - 3e236a3
PASSED - {yarpc-java map[server:yarpc-java behavior:json transport:tchannel]} - 21855bd
PASSED - {yarpc-java map[server:yarpc-java behavior:thrift transport:http]} - 87a346e
PASSED - {yarpc-java map[behavior:thrift transport:tchannel server:yarpc-java]} - c2ed465
```

@abhinav @junchaowu you guys should be able to work on getting Go <-> Java working now.

@kriskowal and I will work on this node issue out of band.

File: Makefile [hunk 1]
```diff
@@ -2,7 +2,8 @@
 
 crossdock: clean
 	docker-compose build yarpc-java
-	docker-compose run crossdock
+	docker-compose run crossdock || \
+		(docker-compose ps -q | xargs -n1 -I% bash -c 'docker logs %; echo --- logs for % ---')
```
Comment (breerly) [inline Makefile]: @kriskowal this should be helpful :)

**Comment by breerly** [comment]
Nice! Go and Java are running successfully!

**Comment by junchaowu** [comment]
:smile:

**Comment by breerly** [comment]
@junchaowu @abhinav @kriskowal added node back, and it works!

:balloon: :balloon: :balloon:

**Comment by breerly** [comment]
:shipit:

**Pull request merged**